frame/vesting/src/vesting_info.rs

```diff
@@ -0,0 +1,122 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2017-2021 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Module to enforce private fields on `VestingInfo`.
+
+use super::*;
+use codec::MaxEncodedLen;
+
+/// Struct to encode the vesting schedule of an individual account.
+#[derive(Encode, Decode, Copy, Clone, PartialEq, Eq, RuntimeDebug, MaxEncodedLen)]
+pub struct VestingInfo<Balance, BlockNumber> {
+	/// Locked amount at genesis.
+	locked: Balance,
+	/// Amount that gets unlocked every block after `starting_block`.
+	per_block: Balance,
+	/// Starting block for unlocking(vesting).
+	starting_block: BlockNumber,
+}
+
+impl<Balance, BlockNumber> VestingInfo<Balance, BlockNumber>
+where
+	Balance: AtLeast32BitUnsigned + Copy,
+	BlockNumber: AtLeast32BitUnsigned + Copy + Bounded,
+{
+	/// Instantiate a new `VestingInfo`.
+	pub fn new<T: Config>(
+		locked: Balance,
+		per_block: Balance,
+		starting_block: BlockNumber,
+	) -> VestingInfo<Balance, BlockNumber> {
+		VestingInfo { locked, per_block, starting_block }
+	}
+
+	/// Validate parameters for `VestingInfo`. Note that this does not check
+	/// against `MinVestedTransfer` or the current block.
+	pub fn validate<BlockNumberToBalance: Convert<BlockNumber, Balance>, T: Config>(
```

Comment:
```suggestion
	pub fn validate<BlockNumberToBalance: Convert<BlockNumber, Balance>>(
```

Reply:
9f1a90c34369654d5a6852b930ae1a9f45796fe8

We need the `Config` for the error type (unless there is a way around this), but we actually can get rid of `BlockNumberToBalance`

Reply:
Have this function return `()` for error, and then when you call this function `.map_err(|_| Error::<T>::InvalidScheduleParams)`

Reply:
Or i think you can just have it return a `DispatchError`

Reply:
With `DispatchError` don't I still need the `Config` trait so I can do `Error::<T>::InvalidScheduleParams`?

I like the `.map_err` approach though

Reply:
DispatchError is probably the right choice

Reply:
I Implemented with `.map_err` here before I saw the comment: 7a05c7162f6bcc650b131b9b6fd5eec10015dcc6

Looking at `DispatchError`, if I want to use [Module error](https://docs.rs/frame-support/3.0.0/frame_support/dispatch/enum.DispatchError.html#variant.Module), I think I need `Config` to  because the error and the module it belongs too are generic over `Config`. Gonna take a closer look and see if I can do something tho with DispatchError and not Config

Reply:
yeah I can't figure out how to make it work. I added a non bounded type param `T` to validate so I can do `Error::<T>..`, since the `Error` type param is not explicitly bounded, but given the following error I don't think it will work : 
```
error[E0277]: the trait bound `T: pallet::Config` is not satisfied
  --> frame/vesting/src/vesting_info.rs:50:3
   |
50 |         ensure!(!self.locked.is_zero() && !self.raw_per_block().is_zero(), Error::<T>::InvalidScheduleParams);
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `pallet::Config` is not implemented for `T`
   |
   = note: required because of the requirements on the impl of `From<pallet::Error<T>>` for `sp_runtime::DispatchError`
   = note: required because of the requirements on the impl of `Into<sp_runtime::DispatchError>` for `pallet::Error<T>`
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider restricting type parameter `T`
   |
49 |     pub fn validate<T: pallet::Config>(&self) -> Result<(), DispatchError> {
```

I can't find any From or Into impl related too `From<pallet::Error<T>>` for `sp_runtime::DispatchError`, but I assume that bounds `T: Config`

Reply:
changed return type to bool and fn ident to `is_valid` 81a29af